File: python-workers/utils/redact.py
import re

def redact(data):
    if isinstance(data, Exception):
        return {
            'type': type(data).__name__,
            'message': redact(str(data)),
            'args': redact(data.args)
        }
    
    
    if isinstance(data, str):
        result = data
        result = re.sub(r'password["\s:]*["\']?([^"\']+)["\']?', 'password: [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'token["\s:]*["\']?([^"\']+)["\']?', 'token: [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'apikey["\s:]*["\']?([^"\']+)["\']?', 'apikey: [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'secret["\s:]*["\']?([^"\']+)["\']?', 'secret: [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'authorization["\s:]*["\']?Bearer\s+([^"\']+)["\']?', 'authorization: Bearer [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'Bearer\s+[^\s]+', 'Bearer [REDACTED]', result, flags=re.IGNORECASE)
        result = re.sub(r'(mongodb|postgres|mysql|mssql)://[^\s]+', '[DATABASE_URL_REDACTED]', result, flags=re.IGNORECASE)
        return result
    
    if isinstance(data, dict):
        redacted = {}
        sensitive_keys = [
            'password', 'pwd', 'secret', 'token', 'apikey', 'api_key',
            'authorization', 'bearer', 'privatekey', 'private_key',
            'refresh_token', 'refreshtoken', 'access_token', 'accesstoken'
        ]
        
        for key, value in data.items():
            if any(k in key.lower() for k in sensitive_keys):
                redacted[key] = '[REDACTED]'
            elif isinstance(value, dict) or isinstance(value, list):
                redacted[key] = redact(value)
            elif isinstance(value, str):
                redacted[key] = redact(value)
            else:
                redacted[key] = value
        return redacted
    
    if isinstance(data, list):
        return [redact(item) for item in data]
    
    return data

File: qqtech9-backend/src/utils/redact.js
function redact(data) {
  if (typeof data === 'string') {
    return data
        .replace(/email["\s:]*["\']?([^"\']+)["\']?/gi, 'email: [REDACTED]')
        .replace(/password["\s:]*["\']?([^"\']+)["\']?/gi, 'password: [REDACTED]')
        .replace(/token["\s:]*["\']?([^"\']+)["\']?/gi, 'token: [REDACTED]')
        .replace(/phone["\s:]*["\']?([^"\']+)["\']?/gi, 'phone: [REDACTED]')
        .replace(/apikey["\s:]*["\']?([^"\']+)["\']?/gi, 'apikey: [REDACTED]')
        .replace(/secret["\s:]*["\']?([^"\']+)["\']?/gi, 'secret: [REDACTED]')
        .replace(/authorization["\s:]*["\']?Bearer\s+([^"\']+)["\']?/gi, 'authorization: Bearer [REDACTED]')
        .replace(/Bearer\s+[^\s]+/gi, 'Bearer [REDACTED]')
        .replace(/(mongodb|postgres|mysql|mssql):\/\/[^\s]+/gi, '[DATABASE_URL_REDACTED]');
    }

  if (typeof data === 'object' && data !== null) {
    if (Array.isArray(data)) {
        return data.map(item => redact(item));
    }

    const redacted = {};
    for (const [key, value] of Object.entries(data)) {
        if (['email', 'password', 'pwd', 'secret', 'token', 'phone', 'apiKey', 'api_key', 'authorization', 'bearer', 'privateKey', 'refresh_token', 'accessToken'].some(k => key.toLowerCase().includes(k))) {
        redacted[key] = '[REDACTED]';
        } else if (typeof value === 'object') {
        redacted[key] = redact(value);
        } else if (typeof value === 'string') {
        redacted[key] = redact(value);
        } else {
        redacted[key] = value;
        }
    }
    return redacted;
  }

  return data;
}

export { redact };
